### ISSUE TITLE
Intermittent "Uncaught RuntimeError: null function or function signature mismatch at __fwritex"

### ISSUE THREAD

**Issue by liquidaty** [emscripten-core/emscripten#17953]
Please include the following in your bug report:

**Version of emscripten/emsdk:**
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.20 (5d878c99921ec247d34fb26a20b5a13d60d69e93)
clang version 16.0.0 (https://github.com/llvm/llvm-project 75767a0f9a926641edbef08e31ec2148ff45da67)
Target: wasm32-unknown-emscripten
Thread model: posix

Am calling an emscripten-compiled executable from browser (i.e. calling main() and passing it some args). For a certain command, if I run that command many times in a row, it always succeeds the first time, and then it fails with the above error about 50% of the time on subsequent calls. After it fails, the next time it might succeed again, or might fail again. I'm unfortunately unable to share the rather large codebase I'm working with.

This issue does not occur when the same code is compiled to run natively (e.g. with gcc or clang). Running as a native binary, valgrind shows no issues, leaks has no issues, -fsanitize=address has no issues.

Using emcc, the issue is the same with -00, -O2 and -O3. Tried compiling with -fsanitize=address, but when the binary is run from the browser it just hangs and the callback it should invoke prior to exit is never invoked (nor is any issue log output). Tried with -s EXIT_RUNTIME and I get "user callback triggered after runtime exited". 

Whatever the problem is, it very well may be coming from a flaw in our code; the trace always shows it occurring in the same place (though we call fwrite in many other places). Nonetheless, it seems to be happening with a perfectly valid fwrite() call, does not occur under other compilers / environments, and does not show up on any diagnostics we have run. At this point I'm running out of ideas as to how to identify what that issue could be. Any suggestions as to how to diagnose where this issue could be coming from or possible workarounds?

**Comment by liquidaty** [comment]
Updates:
* If we turn on a verbose switch in our code to add a lot of fprintf(stderr, ...) statements at each step, the issue seems to go away, or at least become far less likely to occur
* once in a blue moon the error is "Uncaught RuntimeError: table index is out of bounds  at __fwritex"
* when compiled with -fsanitize=undefined, the wasm module fails to load (local count too large @+nnnnnnn)

**Comment by sbc100** [comment]
Can you share the full link command you are using?  (so we can see all your settings)

**Comment by sbc100** [comment]
I think the `-sEXIT_RUNTIME` thing might be a clue.  Can you get a JS backtrace at that point where the `"user callback triggered after runtime exited"` is reported?  Why is triggering that?

**Comment by liquidaty** [comment]
Re EXIT_RUNTIME, unfortunately after trying to add further diagnostic code that error is no longer appearing. Instead, I get: "Uncaught InvalidStateError: Failed to execute 'transaction' on 'IDBDatabase': The database connection is closing" in my JS which is trying an FS sync immediately after the wasm completes

The code does use emscripten fetch and executes a callback when the fetch completes. I'm guessing that main is exiting before fetch completes, causing the above issue. Alternatively is there some emscripten-specific `atexit()` function that could be manually called so that in the case of an async operation, I can try that approach to clean up when that completes?

**Comment by liquidaty** [comment]
I've found and fixed the bug in our code: fwrite was being called (to write a tiny amount of data) on a FILE * that had just been closed with `fclose`. While it's surprising that the error did not cause any issue running under other environments / os's (macOS and windows), I also wonder:

1. why was the error "null function or function signature mismatch" and not something else like "invalid IO operation"
2. why was it an intermittent error-- why wouldn't it always occur?

Anyway, I am guessing that since the FILE * was already closed by fclose(), then fwrite was effectively being given garbage memory and who knows what it was actually writing to, so not sure I can expect a proper error message.

@sbc100 thank you, appreciate the suggestions which were indirectly helpful in resolving
